distribution/tools/launchers/src/main/java/org/elasticsearch/tools/launchers/MachineDependentHeap.java

```diff
@@ -0,0 +1,247 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.tools.launchers;
+
+import org.yaml.snakeyaml.Yaml;
+import org.yaml.snakeyaml.error.YAMLException;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+
+import static java.lang.Math.max;
+import static java.lang.Math.min;
+
+/**
+ * Determines optimal default heap settings based on available system memory and assigned node roles.
+ */
+public final class MachineDependentHeap {
+    private static final long GB = 1024L * 1024L * 1024L; // 1GB
+    private static final long MAX_HEAP_SIZE = GB * 31; // 31GB
+    private static final long MAX_ML_HEAP_SIZE = GB * 2; // 2GB
+    private static final long MIN_HEAP_SIZE = 1024 * 1024 * 128; // 128MB
+    private static final int DEFAULT_HEAP_SIZE_MB = 1024;
+    private static final String ELASTICSEARCH_YML = "elasticsearch.yml";
+
+    private final SystemMemoryInfo systemMemoryInfo;
+
+    public MachineDependentHeap(SystemMemoryInfo systemMemoryInfo) {
+        this.systemMemoryInfo = systemMemoryInfo;
+    }
+
+    /**
+     * Calculate heap options.
+     *
+     * @param configDir path to config directory
+     * @param userDefinedJvmOptions JVM arguments provided by the user
+     * @return final heap options, or an empty collection if user provided heap options are to be used
+     * @throws IOException if unable to load elasticsearch.yml
+     */
+    public List<String> determineHeapSettings(Path configDir, List<String> userDefinedJvmOptions) throws IOException {
+        if (userDefinedJvmOptions.stream().anyMatch(s -> s.startsWith("-Xms") || s.startsWith("-Xmx"))) {
```

Comment:
I don't think check is sufficient. A user can also set `-XX:MaxHeapSize` and `-XX:MinHeapSize` for which `-Xmx` and `-Xms` are synonyms.

Note that in `JvmErgonomics#choose` we start up a second JVM and extract whether or not a setting was passed in on the command line. We might want to reuse that here and then check if `MinHeapSize` or `MaxHeapSize` are passed on the command line. This might be more reliable?

Reply:
I've moved the method to find the final options so it is reused here.